### ISSUE TITLE
Web IRC support

### ISSUE THREAD

**Issue by mweinelt** [matrix-org/matrix-appservice-irc#153]
Please implement the Web IRC protocol. It enables opers to properly react towards abusive behaviour, without needing to ban the entire appservice.

https://kiwiirc.com/docs/webirc

**Comment by Kegsay** [comment]
Not sure this will work for us. WebIRC assumes that the real user is connecting to the intermediary, which then goes through to the IRCd. This is not true in Matrix. We get HTTP requests from the HS, not the user. This means the bridge doesn't know the IP address of the client.

**Comment by itsrachelfish** [comment]
Would it be possible for the homeserver specify `X_FORWARDED_FOR` HTTP headers? If this isn't currently an option in synapse, maybe it could at least be created as a feature request?

**Comment by turt2live** [comment]
> no, it can also be dynamic and I am not sure the appservice knows that either and I don't know how it handles multiple IPs (IPv4/IPv6) aed what if the one it used went down?

There's also the problem of the appservice transactions likely containing multiple user's messages, therefore the IP is relatively irrelevant. Similarly, synapse may not be aware of a remote user's IP.

**Comment by Mikaela** [comment]
How difficult/timeconsuming would this be to implement or would it be better to skip/close this in favour of S2S protocol (https://github.com/matrix-org/matrix-appservice-irc/issues/329)?

**Comment by Kegsay** [comment]
S2S would probably be a better solution if I'm honest. Both options would be time consuming to implement, but S2S has other benefits so would probably be preferable.

**Comment by rubinlinux** [comment]
We solved a similar problem on AfterNET with regards to ZNC bouncers, where we don't necessarily know the users IP. So we 'make one up' in reserved space by hashing their username. That way the same user always gets the same IP and if they abuse, any glines/klines set on that IP wont affect everyone from the same bouncer. That way on the IRC side bans and things all 'just work' as expected.

**Comment by mweinelt** [comment]
FWIW: If you're setup can ban based on Gecos/Realname, that's where the matrix id is visible.

On Charybdis for example: `/mode #example +b $r:user@example.com`.

**Comment by Mikaela** [comment]
At Ergo IRCd there are thoughts of a separate command for communicating bridge identifiers across https://github.com/ergochat/ergo/issues/1724#issuecomment-873842622, which I think could possibly replace WEBIRC for this use case (if it also got widely implemented) and resolve the worry on real name being depended upon for managing abuse https://github.com/matrix-org/matrix-appservice-irc/issues/723.

**Comment by mogad0n** [comment]
yeah dropped in here to link the Ergo comments but it would indeed be nice to have proper standardized way of handling puppet based bridges.
